File: algorithms/code/leetcode/lc034_search_for_a_range/LC034SearchForARange.java
package leetcode.lc034_search_for_a_range;

/**
 * https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
 * #medium
 */
public final class LC034SearchForARange {
    private int bsInfimum(final int[] nums, final int target) {
        int a = 0;
        int z = nums.length - 1;
        while (a < z) {
            int m = (a + z) >>> 1;
            if (nums[m] < target) {
                a = m + 1;
            }
            if (nums[m] == target) {
                z = m;
            }
            if (nums[m] > target) {
                z = m - 1;
            }
        }
        if (/* a == z && */nums[a] == target) {
            return a;
        } else {
            return -1;
        }
    }

    private int bsSupremum(final int[] nums, final int target) {
        int a = 0;
        int z = nums.length - 1;
        while (a < z) {
            int m = (1 + a + z) >>> 1;
            if (nums[m] < target) {
                a = m + 1;
            }
            if (nums[m] == target) {
                a = m;
            }
            if (nums[m] > target) {
                z = m - 1;
            }
        }
        if (a == z && nums[a] == target) {
            return a;
        } else {
            return -1;
        }
    }

    public int[] searchRange(final int[] nums, final int target) {
        if (nums == null || nums.length == 0) {
            return new int[] { -1, -1 };
        }
        return new int[] { bsInfimum(nums, target), bsSupremum(nums, target) };
    }
}

File: algorithms/code/leetcode/lc034_search_for_a_range/lc034_search_for_a_range.rs
// https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
// #medium

fn bs_infimum(nums: &[i32], target: i32) -> i32 {
    let mut a = 0;
    let mut z = nums.len() as i32 - 1;
    while a < z {
        let m = (a + z) >> 1;
        match nums[m as usize].cmp(&target) {
            std::cmp::Ordering::Less => a = m + 1,
            std::cmp::Ordering::Equal => z = m,
            std::cmp::Ordering::Greater => z = m - 1,
        };
    }
    if a == z && nums[a as usize] == target {
        return a;
    }
    -1
}

fn bs_supremum(nums: &[i32], target: i32) -> i32 {
    let mut a = 0;
    let mut z = nums.len() as i32 - 1;
    while a < z {
        let m = (a + z + 1) >> 1;
        match nums[m as usize].cmp(&target) {
            std::cmp::Ordering::Less => a = m + 1,
            std::cmp::Ordering::Equal => a = m,
            std::cmp::Ordering::Greater => z = m - 1,
        };
    }
    if a == z && nums[a as usize] == target {
        return a;
    }
    -1
}

pub fn search_range(nums: Vec<i32>, target: i32) -> Vec<i32> {
    if nums.is_empty() {
        return vec![-1, -1];
    }
    let infimum = bs_infimum(&nums, target);
    let supremum = bs_supremum(&nums, target);
    vec![infimum, supremum]
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example_1() {
        let nums = vec![5, 7, 7, 8, 8, 10];
        assert_eq!(vec![3, 4], search_range(nums, 8));
    }

    #[test]
    fn test_other() {
        let nums = vec![5, 7, 7, 8, 8, 10];
        assert_eq!(vec![-1, -1], search_range(nums, 6));
    }

    #[test]
    fn test_another() {
        let nums = vec![2, 2];
        assert_eq!(vec![-1, -1], search_range(nums, 3));
    }

    #[test]
    fn test_nothing() {
        assert_eq!(vec![-1, -1], search_range(Vec::new(), 3));
    }
}
